extensions/amp-byside-content/0.1/amp-byside-content.js

```diff
@@ -0,0 +1,346 @@
+    this.webcareId_ = '';
+
+    /** @private {string}  */
+    this.channel_ = '';
+
+    /** @private {string}  */
+    this.lang_ = '';
+
+    /** @private {string}  */
+    this.fid_ = '';
+
+    /** @private {string}  */
+    this.label_ = '';
+
+    /** @private {string}  */
+    this.origin_ = '';
+
+    /** @private {string}  */
+    this.baseUrl_ = '';
+
+    /** @const {function()} */
+    this.boundUpdateSize_ =
+		utils.debounce(this.updateSize_.bind(this), 100);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * @param {boolean=} onLayout
+   * @override
+   */
+  preconnectCallback(onLayout) {
+    if (this.iframeSrc_) {
+      this.preconnect.url(this.iframeSrc_, onLayout);
+    }
+  }
+
+  /** @override */
+  buildCallback() {
+    this.webcareId_ = user().assert(
+        this.element.getAttribute('data-webcare-id'),
+        'The data-webcare-id attribute is required for <' + TAG_ + '> %s',
+        this.element);
+
+    this.label_ = user().assert(
+    	this.element.getAttribute('data-label'),
+        'The data-label attribute is required for <' + TAG_ + '> %s',
+        this.element);
+
+    this.webcareZone_ = (this.element.getAttribute('data-webcare-zone') ||
+		DEFAULT_WEBCARE_ZONE_);
+    this.channel_ = (this.element.getAttribute('data-channel') || '');
+    this.lang_ = (this.element.getAttribute('data-lang') || DEFAULT_LANG_);
+    this.fid_ = (this.element.getAttribute('data-fid') || '');
+
+    this.origin_ = this.composeOrigin_();
+    this.baseUrl_ = this.origin_ + '/BWA' +
+		encodeURIComponent(this.webcareId_) + '/amp/';
+  }
+
+  /** @override */
+  createPlaceholderCallback() {
+    const placeholder = this.win.document.createElement('div');
+    placeholder.setAttribute('placeholder', '');
+    placeholder.appendChild(this.createBySideLoader_());
+
+    this.applyFillContent(placeholder);
+
+    return placeholder;
+  }
+
+  /** @override */
+  layoutCallback() {
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    this.iframe_ = iframe;
+
+    iframe.name = 'amp_byside_content_iframe' + iframeCount_++;
+    iframe.setAttribute('scrolling', 'no');
+    iframe.setAttribute('frameborder', '0');
+    iframe.setAttribute('allowtransparency', 'true');
+    iframe.setAttribute('allowfullscreen', 'true');
+    iframe.setAttribute('title', this.element.getAttribute('title') || '');
+
+    setStyles(iframe, {
+      'opacity': 0,
+    });
+
+    this.element.appendChild(this.getOverflowElement_());
+    this.applyFillContent(iframe);
+
+    return this.composeSrcUrl_().then(src => {
+      this.iframeSrc_ = assertHttpsUrl(src, this.element, this.getName_());
+      iframe.src = this.iframeSrc_;
+
+	  const unlisten = listenFor(iframe, 'embed-size', this.boundUpdateSize_);
+	  this.unlisteners_.push(unlisten);
+
+	  this.element.appendChild(iframe);
+	  this.iframePromise_ = this.loadPromise(iframe).then(() => {
```

Comment:
loadPromise is not being returns meaning layoutCallback indicates it is done much earlier than it should. let's refactor a bit here to make it ready better, I suggest:

```
// pseudo-code 
return compose().then({
  ...
 return this.iframePromise_ = this.loadPromise(iframe)
}).then( () => {
 return this.getVsync().mutate ....
});

```